x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

```diff
@@ -5527,4 +5527,39 @@ private Literal nullOf(DataType dataType) {
     public static EsRelation relation() {
         return new EsRelation(EMPTY, new EsIndex(randomAlphaOfLength(8), emptyMap()), randomFrom(IndexMode.values()), randomBoolean());
     }
+
+    public void testReferencePlanSimple() {
+        var plan = optimizedPlan("""
+            from test
+            """);
+
+        ReferencePlan expected = ReferencePlan.limit()
+            .withChild(ReferencePlan.relation());
+        assertTrue(expected.matches(plan));
+
+        expected = ReferencePlan.limit()
+            .withExpression(new Literal(EMPTY, 1000, INTEGER))
+            .withChild(ReferencePlan.relation());
+        assertTrue(expected.matches(plan));
+    }
+
+    public void testReferencePlanTopN() {
+        var plan = optimizedPlan("""
+            from test
+            | sort salary
+            | limit 1
+            """);
+
+        ReferencePlan expected = ReferencePlan.topN();
+        assertTrue(expected.matches(plan));
+
+        expected = ReferencePlan.topN().withChild(ReferencePlan.relation());
+        assertTrue(expected.matches(plan));
+
+        expected = ReferencePlan.topN().withExpression(new Literal(EMPTY, 1, INTEGER));
+        assertTrue(expected.matches(plan));
+
+        expected = ReferencePlan.topN().withChild(ReferencePlan.relation()).withExpression(new Literal(EMPTY, 1, INTEGER));
+        assertTrue(expected.matches(plan));
```

Comment:
Main idea of this prototype: define only the relevant part of the expected logical plan; anything not explicitly defined will be skipped while matching.